pbj-core/pbj-runtime/src/main/java/com/hedera/pbj/runtime/io/stream/ReadableStreamingData.java

```diff
@@ -130,8 +132,97 @@ public long skip(final long n) {
             long numSkipped = in.skip(clamped);
             position += numSkipped;
             return numSkipped;
-        } catch (IOException e) {
+        } catch (final IOException e) {
+            throw new DataAccessException(e);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public long readBytes(@NonNull final byte[] dst, final int offset, final int maxLength) {
+        if (maxLength < 0) {
+            throw new IllegalArgumentException("Negative maxLength not allowed");
+        }
+        if (!hasRemaining()) {
+            return 0;
+        }
+        try {
+            int bytesRead = in.read(dst, offset, maxLength);
+            position += bytesRead;
+            if (bytesRead < maxLength) {
+                final int nextByte = tryReadNextByte();
+                if (nextByte != -1) {
+                    dst[offset + bytesRead] = (byte) nextByte;
+                    bytesRead++;
+                }
+            }
+            return bytesRead;
+        } catch (final IOException e) {
             throw new DataAccessException(e);
         }
     }
+
+    @Override
+    public long readBytes(@NonNull final ByteBuffer dst) {
+        if (!dst.hasArray()) {
+            return ReadableSequentialData.super.readBytes(dst);
+        }
+        if (!hasRemaining()) {
+            return 0;
+        }
+        final byte[] dstArr = dst.array();
+        final int dstArrOffset = dst.arrayOffset();
+        final int dstPos = dst.position();
+        final long len = Math.min(remaining(), dst.remaining());
+        try {
+            int bytesRead = in.read(dstArr, dstPos + dstArrOffset, Math.toIntExact(len));
+            position += bytesRead;
+            if (bytesRead < len) {
+                final int nextByte = tryReadNextByte();
+                if (nextByte != -1) {
+                    dst.put((byte) nextByte);
+                    bytesRead++;
+                }
+            }
+            return bytesRead;
+        } catch (final IOException e) {
+            throw new DataAccessException(e);
+        }
+    }
+
+    @Override
+    public long readBytes(@NonNull final BufferedData dst) {
+        final long len = Math.min(remaining(), dst.remaining());
+        int bytesRead = dst.writeBytes(in, Math.toIntExact(len));
```

Comment:
We should have an alternative `writeBytes` (it is very odd that we use `writeBytes` to _read_ data), or better a `readNBytes`, method on BufferedData that uses `readNBytes` on the input stream so that we do not need these single-byte-reading workarounds for when we need to block efficiently until the full requested length is read (or EOS is reached).

Reply:
See my comment above on `readNBytes()`. It's a design question, what behavior we'd expect from these `readBytes()` and similar methods.

Regarding reads vs writes, I agree it may look odd at a glance. After a while, it starts feeling quite natural, though. Reading from this to dst is the same as writing this to dst. Similar, writing src to this is the same as reading from src to this.

Now let's take a look at `BufferedData.writeBytes(InputStream)`. It's still single-byte writing code, unless the buffered data object is backed by a byte array. Instead of using `dst.writeBytes()` here, I could check if dst has a byte array and use `in.read(array)` as a shortcut. Unfortunately (or fortunately, actually), `BufferedData` provides no API to check if it's direct of byte array based, and no API to get access to the underlying byte array. This is why `dst.writeBytes()` is used instead, and let `BufferedData` subclasses handle it internally.

Reply:
Switched to `readNBytes()` here